### PR TITLE
Support required paths for indexes

### PR DESCRIPTION
**Pull request by gberdzenishvili**

### CODE REVIEW

**Comment by chris-armstrong** [comment]
I'm wondering if this is the correct approach.

Partition keys and Sort Keys are used slightly differently depending on the desired pattern.

There are some competing patterns here:
* Sparse indexing - items are only added to a GSI if both the partition key and sort key are specified
* Hierarchical indexing - you can retrieve items at different levels of a hierarchy if you specify a partial sort key e.g. if you have `pk = <country>` and `sk = <state>#<suburb>#<street>`, you can retrieve all the addresses for a suburb with a `begins_with(sk, <state>#<suburb>#)`, or to retrieve all the entries for a country, you could specify just the partition key (assuming something is put into the sort key). (Partition key must always be specified for a `PK` but sort key is optional / partial)

For this reason, I think to support both sparse indexing and hierarchical index we sort of need to:
1. Allow for missing partition key paths to not index items (with the opposite being - if one of the key paths is missing from a replace or only some of them are specified in an update, throw an error)
2. Allow for missing sort key paths to not index items (at the moment, we just write the collection name to the index so the value still appears - we may want this behaviour to be configurable for a collection)
3. Throw an error if one of the 'required' partition key paths is missing (this conflicts with (1))
4. Throw an error if one of the 'required' sort keys paths is missing

We also want to consider that during an update, we only need to check for all the required key paths to appear in the same 'key' i.e. if you update any key path in a partition key, then any required key paths in the partition key need to be specified, but you don't need to specify any of the required sort key paths if you are not trying to update any of them. Similarly, if you specify key paths which are not part of the partition or sort key, you don't need to do required checks.

This seems rather complicated and we may need to map out all the scenarios to support all of this correctly.

**Comment by chris-armstrong** [comment]
For later context: agreed to continue with the original approach (introducing a new `required` field for specifying required paths), but ensure that when doing updates, all paths for a key (partition or sort) are specified, even if they are optional, so that the intention of updates is clear (no partial key updates).

File: src/base/util.ts [hunk 1]
```diff
@@ -182,14 +186,28 @@ export const constructKeyValue = <DocumentType extends DocumentWithId>(
         { keyPath: valuePath, collection: collectionName }
       );
     }
+    const isRequired = requiredPaths?.find((r) => isEqual(r, valuePath));
+    // by this line it can only be undefined or string
+    if (isRequired && !extractedValue) {
```
Comment (chris-armstrong) [inline src/base/util.ts]: Shouldn't this allow for required but blank values (i.e. empty string)?

**Review by chris-armstrong — APPROVED** [review]
I think this gets us where we need

File: src/base/util.ts (hunk 1 above)
Comment (gberdzenishvili) [reply src/base/util.ts]: yeah, I was hesitant about the definition of required, should we allow nulls as well then to align with json schema definition of required? But then it means evrey index path is required on update (which is probably fine)

File: src/operations/update_by_id.test.ts [hunk 2]
```diff
@@ -606,32 +606,6 @@ describe('mapAccessPatterns', () => {
     ).toThrow(InvalidIndexedFieldValueException);
   });
 
-  it.each([null, undefined, ''])(
```
Comment (chris-armstrong) [inline src/operations/update_by_id.test.ts]: We should have a default handling for `null` values in an index - ideally when we convert them they become empty strings. I'd like to see a test for this

Comment (gberdzenishvili) [reply src/operations/update_by_id.test.ts]: yep, there is a test [here](https://github.com/gberdzenishvili/dynaglue/blob/05a2412483791a14688961bf1a7c6a4c5ba2ab4c/src/operations/update_by_id.test.ts#L875) and e2e test [here](https://github.com/gberdzenishvili/dynaglue/blob/6bf45b73caffbdec94d1a17e8f285a7f73d04f0f/src/operations/e2e.test.ts#L275)

**Pull request merged**